internal/jobworker/cgroup/service.go

```diff
@@ -0,0 +1,270 @@
+}
+
+// CreateCgroup creates a new Service Cgroup. CgroupOptions may be specified to
+// configure the Cgroup. On success, the created Cgroup is returned to the
+// caller.
+func (s Service) CreateCgroup(options ...CgroupOption) (*Cgroup, error) {
+	id := uuid.New()
+	cgroup := &Cgroup{
+		ID:      id,
+		service: s,
+		path:    path.Join(s.path, id.String()),
+	}
+	for _, option := range options {
+		option(cgroup)
+	}
+
+	if err := cgroup.create(); err != nil {
+		return nil, errors.WithStack(err)
+	}
+
+	return cgroup, nil
+}
+
+// PlaceInCgroup places the pid in the Service cgroup specified.
+func (s Service) PlaceInCgroup(cgroup Cgroup, pid int) error {
+	return errors.WithStack(cgroup.placePID(pid))
+}
+
+// RemoveCgroup removes the jobworker cgroup uniquely identified by the
+// specified id.
+func (s Service) RemoveCgroup(id uuid.UUID) error {
+	cgroup := Cgroup{ID: id, service: s, path: path.Join(s.path, id.String())}
+
+	return errors.WithStack(cgroup.remove())
+}
+
+// Cleanup removes all jobworker Service resources. Whenever a Service instance
+// is used, Cleanup should always be called before application close.
+func (s Service) Cleanup() error {
+	if err := s.cleanup(); err != nil {
+		return err
+	}
+
+	if err := s.unmount(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// placeInRootCgroup moves the pids into the root cgroup.
+func (s Service) placeInRootCgroup(pids []int) error {
+	file := path.Join(s.mountPath, cgroupProcs)
+	fd, err := os.OpenFile(file, os.O_WRONLY, fileMode)
+	if err != nil {
+		return errors.WithStack(err)
+	}
+	defer fd.Close()
+
+	for _, pid := range pids {
+		if _, err := fd.WriteString(strconv.Itoa(pid)); err != nil {
+			return errors.WithStack(err)
+		}
+	}
+
+	return nil
+}
+
+// mount setups the cgroup2 filesystem and creates a cgroup dedicated to
+// jobworker cgroups.
+func (s Service) mount() error {
+	// Ensure path to cgroup2 mount point exists.
+	if err := os.MkdirAll(s.mountPath, fileMode); err != nil {
+		return errors.Wrapf(err, "path: %s", s.mountPath)
+	}
+
+	// If the mount path does not exist or has no entries, mount the cgroup
+	// filesystem, and make base directory for jobworker cgroups.
+	entries, err := os.ReadDir(s.mountPath)
+	if err != nil || len(entries) == 0 {
+		goto mount
+	}
+
+	// cgroup2 filesystem is mounted, ensure jobworker base directory exists and
+	// return.
+	if err := os.MkdirAll(s.path, fileMode); err != nil {
+		return errors.WithStack(err)
+	}
+	return nil
+
+mount:
+	if err := unix.Mount("none", s.mountPath, "cgroup2", 0, ""); err != nil {
+		return errors.Wrapf(err, "path: %s", s.mountPath)
+	}
+
+	// create jobworker base directory for jobworker cgroups.
+	if err := os.MkdirAll(s.path, fileMode); err != nil {
+		return errors.WithStack(err)
+	}
+
+	return nil
+}
+
+// cleanup walks the Service base directory, moving all jobworker pids into the
+// root cgroup and removing the each cgroup directory.
+func (s Service) cleanup() error {
```

Comment:
I wouldn't bother with the cleanup. Let's keep the scope small, this is a good chunk of code we don't really need.

Reply:
Question I asked above is relevant here.